presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -139,45 +142,72 @@ private PlanNodeStatsEstimate computeRightJoinStats(
                 rightJoinComplementStats);
     }
 
-    private PlanNodeStatsEstimate computeInnerJoinStats(JoinNode node, PlanNodeStatsEstimate crossJoinStats, Session session, TypeProvider types)
+    private PlanNodeStatsEstimate computeInnerJoinStats(JoinNode node, PlanNodeStatsEstimate crossJoinEstimate, Session session, TypeProvider types)
     {
-        List<EquiJoinClause> equiJoinClauses = node.getCriteria();
+        List<EquiJoinClause> equiJoinCriteria = node.getCriteria();
 
+        if (equiJoinCriteria.isEmpty()) {
+            if (!node.getFilter().isPresent()) {
+                return crossJoinEstimate;
+            }
+            return filterStatsCalculator.filterStats(crossJoinEstimate, node.getFilter().get(), session, types);
+        }
+
+        PlanNodeStatsEstimate equiJoinEstimate = filterByEquiJoinClauses(crossJoinEstimate, node.getCriteria(), session, types);
+
+        if (equiJoinEstimate.isOutputRowCountUnknown()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        if (!node.getFilter().isPresent()) {
+            return equiJoinEstimate;
+        }
+
+        PlanNodeStatsEstimate filteredEquiJoinEstimate = filterStatsCalculator.filterStats(equiJoinEstimate, node.getFilter().get(), session, types);
+
+        if (filteredEquiJoinEstimate.isOutputRowCountUnknown()) {
+            return equiJoinEstimate.mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+        }
+
+        return filteredEquiJoinEstimate;
+    }
+
+    private PlanNodeStatsEstimate filterByEquiJoinClauses(
+            PlanNodeStatsEstimate stats,
+            Collection<EquiJoinClause> clauses,
+            Session session,
+            TypeProvider types)
+    {
+        checkArgument(!clauses.isEmpty(), "clauses is empty");
+        PlanNodeStatsEstimate result = PlanNodeStatsEstimate.unknown();
         // Join equality clauses are usually correlated. Therefore we shouldn't treat each join equality
         // clause separately because stats estimates would be way off. Instead we choose so called
         // "driving clause" which mostly reduces join output rows cardinality and apply UNKNOWN_FILTER_COEFFICIENT
         // for other (auxiliary) clauses.
-        PlanNodeStatsEstimate equiJoinClausesFilteredStats = IntStream.range(0, equiJoinClauses.size())
-                .mapToObj(drivingClauseId -> {
-                    EquiJoinClause drivingClause = equiJoinClauses.get(drivingClauseId);
-                    List<EquiJoinClause> remainingClauses = copyWithout(equiJoinClauses, drivingClauseId);
-                    return filterByEquiJoinClauses(crossJoinStats, drivingClause, remainingClauses, session, types);
-                })
-                .min(comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
```

Comment:
I didn't like the `copyWithout` method. It involves additional copying, that can be avoided by using linked list.

Reply:
You could use `PCollection` instead of `copyWithout` or just simply pass index of element to be excluded. You don't need to modify streamlined processing here

Reply:
> You could use PCollection instead of copyWithout

How is it different than using a LinkedList?

> pass index of element to be excluded

Than the underlying `filterByEquiJoinClauses(crossJoinStats, drivingClause, remainingClauses, session, types)` will be ugly.

Reply:
> How is it different than using a LinkedList?

It doesn't do any actual copy.

> will be ugly.

It's a matter of taste, but I think it's easier to understand than the procedural code here and doesn't completely change flow of the method.